src/setup_payload/AdditionalDataPayloadGenerator.h

```diff
@@ -0,0 +1,81 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file provides a utility to generate Additional Data payload and its members
+ *      (e.g. rotating device id)
+ *
+ */
+
+#pragma once
+#include <core/CHIPError.h>
+#include <system/TLVPacketBufferBackingStore.h>
+#include <type_traits>
+
+#define ROTATING_DEVICE_ID_COUNTER_STR_MAX_LENGTH 10
+#define ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH 16
+#define ROTATING_DEVICE_ID_MAX_LENGTH ROTATING_DEVICE_ID_COUNTER_STR_MAX_LENGTH + ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH
+
+namespace chip {
+namespace RotatingDeviceId {
+static constexpr unsigned kRotatingDeviceIdCounterStringMaxLength = 10;
+static constexpr unsigned kRotatingDeviceIdHashSuffixLength       = 16;
+static constexpr unsigned kRotatingDeviceIdMaxLength = kRotatingDeviceIdCounterStringMaxLength + kRotatingDeviceIdHashSuffixLength;
+static constexpr unsigned kRotatingDeviceIdHexMaxLength = kRotatingDeviceIdMaxLength * 2 + 1;
+} // namespace RotatingDeviceId
+
+enum class AdditionalDataFields : int8_t
+{
+    NotSpecified     = 0x00,
+    RotatingDeviceId = 0x01
+};
+
+inline AdditionalDataFields operator|(AdditionalDataFields lhs, AdditionalDataFields rhs)
+{
+    using T = std::underlying_type_t<AdditionalDataFields>;
+    return static_cast<AdditionalDataFields>(static_cast<T>(lhs) | static_cast<T>(rhs));
+}
+
+inline AdditionalDataFields & operator|=(AdditionalDataFields & lhs, AdditionalDataFields rhs)
+{
+    lhs = lhs | rhs;
+    return lhs;
+}
+
+inline AdditionalDataFields operator&(AdditionalDataFields lhs, AdditionalDataFields rhs)
+{
+    using T = std::underlying_type_t<AdditionalDataFields>;
+    return static_cast<AdditionalDataFields>(static_cast<T>(lhs) & static_cast<T>(rhs));
+}
+
+class AdditionalDataPayloadGenerator
+{
+
+public:
+    AdditionalDataPayloadGenerator() {}
+    // Generate additional data payload (i.e. TLV encoded)
+    CHIP_ERROR generateAdditionalDataPayload(uint16_t lifetimeCounter, const char * serialNumberBuffer,
+                                             size_t serialNumberBufferSize, chip::System::PacketBufferHandle & bufferHandle,
+                                             AdditionalDataFields additionalDataFields);
+    // Generate Device Rotating ID
+    CHIP_ERROR generateRotatingDeviceId(uint16_t lifetimeCounter, const char * serialNumberBuffer, size_t serialNumberBufferSize,
+                                        char * rotatingDeviceIdBuffer, size_t rotatingDeviceIdBufferSize,
+                                        size_t & rotatingDeviceIdBufferOutputSize);
```

Comment:
Maybe this should be named `rotatingDeviceIdValueOutputSize`?  And document that this is the number of chars making up the actual value, so callers don't guess whether the size includes a null-terminator or not?

Reply:
updated